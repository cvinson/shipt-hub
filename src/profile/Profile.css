.profileContainer {
  display: flex;
  align-items: center;
  justify-content: center;
}

.profile {
  max-width: 300px;
}
