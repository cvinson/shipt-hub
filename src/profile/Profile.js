import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { InfiniteLoader, List } from 'react-virtualized';
import CircularProgress from 'material-ui/CircularProgress';
import Paper from 'material-ui/Paper';
import Follower from '../follower/Follower';
import User from '../user/User';

const PAGE_SIZE = 30;

class Profile extends Component {
  static propTypes = {
    getUser: PropTypes.func.isRequired,
    getUserFollowers: PropTypes.func.isRequired,
    history: PropTypes.object.isRequired,
    username: PropTypes.string.isRequired
  }

  constructor(props) {
    super(props);
    this.state = {
      followers: null,
      user: null
    };
  }

  fetchUserData = async (username) => {
    this.setState({
      followers: await this.props.getUserFollowers(username),
      user: await this.props.getUser(username)
    });
  }

  componentDidMount() {
    return this.fetchUserData(this.props.username);
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.username !== nextProps.username) {
      this.fetchUserData(nextProps.username);
    }
  }

  isRowLoaded = ({ index }) => index < this.state.followers.length;

  loadMoreRows = async ({ startIndex, stopIndex }) => {
    const nextPage = Math.ceil(this.state.followers.length / PAGE_SIZE) + 1;
    const additionalFollowers = await this.props.getUserFollowers(this.props.username, nextPage);
    this.setState({ followers: [...this.state.followers, ...additionalFollowers] });
  }

  rowRenderer = ({ key, index, isScrolling, isVisible, style }) => {
    const follower = this.state.followers[index];
    const onClick = () => this.props.history.push(`/${follower.login}`);
    const onLinkClick = () => window.location = `https://github.com/${follower.login}`;
    return (
      <Follower
        key={key}
        onClick={onClick}
        onLinkClick={onLinkClick}
        style={style}
        {...follower} />
    )
  };

  render() {
    if (!this.state.user) {
      return <CircularProgress size={80} />;
    }

    return (
      <Paper className="profile">
        <User {...this.state.user} />
        <InfiniteLoader
          isRowLoaded={this.isRowLoaded}
          loadMoreRows={this.loadMoreRows}
          minimumBatchSize={PAGE_SIZE}
          rowCount={this.state.user.followers}>
          {({ onRowsRendered, registerChild }) => (
            <List
              height={300}
              width={300}
              onRowsRendered={onRowsRendered}
              ref={registerChild}
              rowCount={this.state.followers.length}
              rowHeight={56}
              rowRenderer={this.rowRenderer} />
          )}
        </InfiniteLoader>
      </Paper>
    );
  }
}

export default Profile;
