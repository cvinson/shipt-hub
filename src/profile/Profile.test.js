import React from 'react';
import PropTypes from 'prop-types';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { MemoryRouter } from 'react-router-dom';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import Profile from './Profile';

const renderOptions = {
  context: { muiTheme: getMuiTheme() },
  childContextTypes: { muiTheme: PropTypes.object }
};

const TestWrapper = ({ getUserProp, getUserFollowersProp, username }) => (
  <MemoryRouter>
    <MuiThemeProvider>
      <Profile
        username={username || 'githubUser'}
        getUser={getUserProp}
        getUserFollowers={getUserFollowersProp}
        history={{ push: [].push }} />
    </MuiThemeProvider>
  </MemoryRouter>
);

describe('<Profile />', () => {
  let getUser, getUserProp, getUserFollowers, getUserFollowersProp;

  beforeEach(() => {
    getUser = Promise.resolve({ login: 'githubUser', followers: 1, avatar_url: 'anAvatarUrl' });
    getUserProp = jest.fn(() => getUser);
    getUserFollowers = Promise.resolve([{ login: 'anotherGithubUser', avatar_url: 'anotherAvatarUrl' }]);
    getUserFollowersProp = jest.fn(() => getUserFollowers);
  });

  it('renders correctly', () => {
    const wrapper = shallow(
      <Profile
        username="githubUser"
        getUser={getUserProp}
        getUserFollowers={getUserFollowersProp}
        history={{}} />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders user info and followers', async () => {
    const wrapper = mount(<TestWrapper getUserProp={getUserProp} getUserFollowersProp={getUserFollowersProp} />);

    await Promise.all([getUser, getUserFollowers]);
    wrapper.update();

    expect(wrapper.text()).toMatch('githubUser');
    expect(wrapper.text()).toMatch('1 Follower');
    expect(wrapper.text()).toMatch('anotherGithubUser');
  });

  it('calls the given user and follower fetch function', async () => {
    const wrapper = mount(<TestWrapper getUserProp={getUserProp} getUserFollowersProp={getUserFollowersProp} />);
    wrapper.setProps({ username: 'githubUser' });
    await Promise.all([getUser, getUserFollowers]);

    expect(getUserProp).toHaveBeenCalledTimes(1);
    expect(getUserFollowersProp).toHaveBeenCalledTimes(1);
  });

  it('calls the given user and follower fetch function again when new username props is passed', async () => {
    const wrapper = mount(<TestWrapper getUserProp={getUserProp} getUserFollowersProp={getUserFollowersProp} />);
    await Promise.all([getUser, getUserFollowers]);
    wrapper.setProps({ username: 'aSecondGithubUser' })
    await Promise.all([getUser, getUserFollowers]);

    expect(getUserProp).toHaveBeenCalledTimes(2);
    expect(getUserProp).toHaveBeenLastCalledWith('aSecondGithubUser');
    expect(getUserFollowersProp).toHaveBeenCalledTimes(2);
    expect(getUserFollowersProp).toHaveBeenLastCalledWith('aSecondGithubUser');
  });

  it('renders error correctly', async () => {
    const wrapper = mount(
      <Profile
        username="githubUser"
        getUser={getUserProp}
        getUserFollowers={getUserFollowersProp}
        history={{}} />,
      renderOptions
    );

    await Promise.all([getUser, getUserFollowers]);
    wrapper.setState({ error: 'rateLimit' });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('sets the appropriate error state when user is not found', async () => {
    getUser = Promise.reject({ message: '404' });
    getUserProp = jest.fn(() => getUser);
    const wrapper = mount(
      <Profile
        username="githubUser"
        getUser={getUserProp}
        getUserFollowers={getUserFollowersProp}
        history={{}} />,
      renderOptions
    );

    try {
      await getUser;
    } catch (err) {
      process.nextTick(() => {
        expect(wrapper.state(['error'])).toEqual('notFound');
      });
    }
  });

  it('sets the appropriate error state when we have been rate limited', async () => {
    getUser = Promise.reject({ message: '403' });
    getUserProp = jest.fn(() => getUser);
    const wrapper = mount(
      <Profile
        username="githubUser"
        getUser={getUserProp}
        getUserFollowers={getUserFollowersProp}
        history={{}} />,
      renderOptions
    );

    try {
      await getUser;
    } catch (err) {
      process.nextTick(() => {
        expect(wrapper.state(['error'])).toEqual('rateLimit');
      });
    }
  });

  describe('loadMoreFollowers()', () => {
    it('loads additional followers and concats them with existing followers', async () => {
      const wrapper = mount(
        <Profile
          username="githubUser"
          getUser={getUserProp}
          getUserFollowers={getUserFollowersProp}
          history={{}} />,
        renderOptions
      );
      wrapper.setState({ followers: [] });
      await wrapper.instance().loadMoreFollowers({ startIndex: 0, stopIndex: 10 });
      expect(wrapper.state(['followers'])).toHaveLength(1);
    })
  });

  afterEach(() => {
    getUserProp.mockClear();
    getUserFollowersProp.mockClear();
  });
});
