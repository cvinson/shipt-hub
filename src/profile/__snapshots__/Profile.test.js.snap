// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile renders correctly 1`] = `
<CircularProgress
  max={100}
  min={0}
  mode="indeterminate"
  size={80}
  thickness={3.5}
  value={0}
/>
`;
